// @getpara/core-sdk/dist/cjs/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ACCOUNT_LINK_CONFLICT: () => ACCOUNT_LINK_CONFLICT,
  EXTERNAL_WALLET_CONNECTION_ONLY_USER_ID: () => EXTERNAL_WALLET_CONNECTION_ONLY_USER_ID,
  LOCAL_STORAGE_AUTH_INFO: () => LOCAL_STORAGE_AUTH_INFO,
  LOCAL_STORAGE_COUNTRY_CODE: () => LOCAL_STORAGE_COUNTRY_CODE,
  LOCAL_STORAGE_CURRENT_WALLET_IDS: () => LOCAL_STORAGE_CURRENT_WALLET_IDS,
  LOCAL_STORAGE_ED25519_WALLETS: () => LOCAL_STORAGE_ED25519_WALLETS,
  LOCAL_STORAGE_EMAIL: () => LOCAL_STORAGE_EMAIL,
  LOCAL_STORAGE_EXTERNAL_WALLETS: () => LOCAL_STORAGE_EXTERNAL_WALLETS,
  LOCAL_STORAGE_EXTERNAL_WALLET_USER_ID: () => LOCAL_STORAGE_EXTERNAL_WALLET_USER_ID,
  LOCAL_STORAGE_FARCASTER_USERNAME: () => LOCAL_STORAGE_FARCASTER_USERNAME,
  LOCAL_STORAGE_PHONE: () => LOCAL_STORAGE_PHONE,
  LOCAL_STORAGE_SESSION_COOKIE: () => LOCAL_STORAGE_SESSION_COOKIE,
  LOCAL_STORAGE_TELEGRAM_USER_ID: () => LOCAL_STORAGE_TELEGRAM_USER_ID,
  LOCAL_STORAGE_USER_ID: () => LOCAL_STORAGE_USER_ID,
  LOCAL_STORAGE_WALLETS: () => LOCAL_STORAGE_WALLETS,
  PARA_CORE_VERSION: () => PARA_CORE_VERSION,
  PARA_PREFIX: () => PARA_PREFIX,
  POLLING_INTERVAL_MS: () => POLLING_INTERVAL_MS,
  POLLING_TIMEOUT_MS: () => POLLING_TIMEOUT_MS,
  PREFIX: () => PREFIX,
  SESSION_STORAGE_LOGIN_ENCRYPTION_KEY_PAIR: () => SESSION_STORAGE_LOGIN_ENCRYPTION_KEY_PAIR,
  SHORT_POLLING_INTERVAL_MS: () => SHORT_POLLING_INTERVAL_MS
});
module.exports = __toCommonJS(constants_exports);
const PARA_CORE_VERSION = "2.0.0-alpha.51";
const PREFIX = "@CAPSULE/";
const PARA_PREFIX = "@PARA/";
const LOCAL_STORAGE_AUTH_INFO = `${PREFIX}authInfo`;
const LOCAL_STORAGE_EMAIL = `${PREFIX}e-mail`;
const LOCAL_STORAGE_PHONE = `${PREFIX}phone`;
const LOCAL_STORAGE_COUNTRY_CODE = `${PREFIX}countryCode`;
const LOCAL_STORAGE_FARCASTER_USERNAME = `${PREFIX}farcasterUsername`;
const LOCAL_STORAGE_TELEGRAM_USER_ID = `${PREFIX}telegramUserId`;
const LOCAL_STORAGE_EXTERNAL_WALLET_USER_ID = `${PREFIX}externalWalletUserId`;
const LOCAL_STORAGE_USER_ID = `${PREFIX}userId`;
const LOCAL_STORAGE_ED25519_WALLETS = `${PREFIX}ed25519Wallets`;
const LOCAL_STORAGE_WALLETS = `${PREFIX}wallets`;
const LOCAL_STORAGE_EXTERNAL_WALLETS = `${PREFIX}externalWallets`;
const LOCAL_STORAGE_CURRENT_WALLET_IDS = `${PREFIX}currentWalletIds`;
const LOCAL_STORAGE_SESSION_COOKIE = `${PREFIX}sessionCookie`;
const SESSION_STORAGE_LOGIN_ENCRYPTION_KEY_PAIR = `${PREFIX}loginEncryptionKeyPair`;
const POLLING_INTERVAL_MS = 2e3;
const SHORT_POLLING_INTERVAL_MS = 1e3;
const POLLING_TIMEOUT_MS = 3e5;
const EXTERNAL_WALLET_CONNECTION_ONLY_USER_ID = "EXTERNAL_WALLET_CONNECTION_ONLY";
const ACCOUNT_LINK_CONFLICT = "Account already linked";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ACCOUNT_LINK_CONFLICT,
  EXTERNAL_WALLET_CONNECTION_ONLY_USER_ID,
  LOCAL_STORAGE_AUTH_INFO,
  LOCAL_STORAGE_COUNTRY_CODE,
  LOCAL_STORAGE_CURRENT_WALLET_IDS,
  LOCAL_STORAGE_ED25519_WALLETS,
  LOCAL_STORAGE_EMAIL,
  LOCAL_STORAGE_EXTERNAL_WALLETS,
  LOCAL_STORAGE_EXTERNAL_WALLET_USER_ID,
  LOCAL_STORAGE_FARCASTER_USERNAME,
  LOCAL_STORAGE_PHONE,
  LOCAL_STORAGE_SESSION_COOKIE,
  LOCAL_STORAGE_TELEGRAM_USER_ID,
  LOCAL_STORAGE_USER_ID,
  LOCAL_STORAGE_WALLETS,
  PARA_CORE_VERSION,
  PARA_PREFIX,
  POLLING_INTERVAL_MS,
  POLLING_TIMEOUT_MS,
  PREFIX,
  SESSION_STORAGE_LOGIN_ENCRYPTION_KEY_PAIR,
  SHORT_POLLING_INTERVAL_MS
});
