// @getpara/core-sdk/dist/cjs/errors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var errors_exports = {};
__export(errors_exports, {
  TransactionReviewDenied: () => TransactionReviewDenied,
  TransactionReviewError: () => TransactionReviewError,
  TransactionReviewTimeout: () => TransactionReviewTimeout
});
module.exports = __toCommonJS(errors_exports);
class TransactionReviewError extends Error {
  constructor(transactionReviewUrl) {
    super("transaction review error");
    this.name = "TransactionReviewError";
    this.transactionReviewUrl = transactionReviewUrl;
  }
}
class TransactionReviewDenied extends Error {
  constructor() {
    super("transaction review has been denied by the user");
    this.name = "TransactionReviewDenied";
  }
}
class TransactionReviewTimeout extends Error {
  constructor(transactionReviewUrl, pendingTransactionId) {
    super("transaction review has timed out");
    this.name = "TransactionReviewTimeout";
    this.transactionReviewUrl = transactionReviewUrl;
    this.pendingTransactionId = pendingTransactionId;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TransactionReviewDenied,
  TransactionReviewError,
  TransactionReviewTimeout
});
