// @getpara/core-sdk/dist/esm/index.js
import "./chunk-7B52C2XE.js";
import { ParaCore } from "./ParaCore.js";
import {
  AuthMethod,
  EmailTheme,
  Network,
  WalletType,
  WalletScheme,
  OnRampAsset,
  OnRampPurchaseType,
  OnRampProvider,
  OnRampPurchaseStatus,
  OAuthMethod,
  NON_ED25519,
  PREGEN_IDENTIFIER_TYPES,
  WALLET_TYPES,
  WALLET_SCHEMES,
  OAUTH_METHODS,
  LINKED_ACCOUNT_TYPES,
  EXTERNAL_WALLET_TYPES,
  EVM_WALLETS,
  SOLANA_WALLETS,
  COSMOS_WALLETS
} from "@getpara/user-management-client";
import {
  OnRampMethod,
  PopupType,
  PregenIdentifierType,
  RecoveryStatus,
  AccountLinkError
} from "./types/index.js";
export * from "./types/coreApi.js";
export * from "./types/events.js";
export * from "./types/config.js";
import { getPortalDomain, entityToWallet, constructUrl, shortenUrl } from "./utils/index.js";
import { PREFIX, PARA_PREFIX } from "./constants.js";
import { distributeNewShare } from "./shares/shareDistribution.js";
import { KeyContainer } from "./shares/KeyContainer.js";
import { getBaseUrl, initClient } from "./external/userManagementClient.js";
import * as mpcComputationClient from "./external/mpcComputationClient.js";
import {
  decryptWithKeyPair,
  decryptWithPrivateKey,
  getAsymmetricKeyPair,
  getPublicKeyHex,
  encryptWithDerivedPublicKey,
  encodePrivateKeyToPemHex,
  getDerivedPrivateKeyAndDecrypt,
  getPublicKeyFromSignature,
  getSHA256HashHex,
  encryptPrivateKey,
  decryptPrivateKey,
  decryptPrivateKeyAndDecryptShare,
  hashPasswordWithSalt,
  encryptPrivateKeyWithPassword,
  decryptPrivateKeyWithPassword,
  publicKeyFromHex
} from "./cryptography/utils.js";
export * from "./external/userManagementClient.js";
export * from "./errors.js";
export * from "./utils/formatting.js";
export * from "./utils/polling.js";
export * from "./utils/phone.js";
import { isWalletSupported } from "./utils/wallet.js";
import { getNetworkPrefix, getOnRampAssets, getOnRampNetworks, toAssetInfoArray } from "./utils/onRamps.js";
import { getPortalBaseURL } from "./utils/url.js";
import { retrieve } from "./transmission/transmissionUtils.js";
const paraVersion = ParaCore.version;
var src_default = ParaCore;
export {
  AccountLinkError,
  AuthMethod,
  COSMOS_WALLETS,
  EVM_WALLETS,
  EXTERNAL_WALLET_TYPES,
  EmailTheme,
  KeyContainer,
  LINKED_ACCOUNT_TYPES,
  NON_ED25519,
  Network,
  OAUTH_METHODS,
  OAuthMethod,
  OnRampAsset,
  OnRampMethod,
  OnRampProvider,
  OnRampPurchaseStatus,
  OnRampPurchaseType,
  PARA_PREFIX as PARA_STORAGE_PREFIX,
  PREGEN_IDENTIFIER_TYPES,
  PopupType,
  PregenIdentifierType,
  RecoveryStatus,
  SOLANA_WALLETS,
  PREFIX as STORAGE_PREFIX,
  WALLET_SCHEMES,
  WALLET_TYPES,
  WalletScheme,
  WalletType,
  constructUrl,
  decryptPrivateKey,
  decryptPrivateKeyAndDecryptShare,
  decryptPrivateKeyWithPassword,
  decryptWithKeyPair,
  decryptWithPrivateKey,
  src_default as default,
  distributeNewShare,
  encodePrivateKeyToPemHex,
  encryptPrivateKey,
  encryptPrivateKeyWithPassword,
  encryptWithDerivedPublicKey,
  entityToWallet,
  getAsymmetricKeyPair,
  getBaseUrl,
  getDerivedPrivateKeyAndDecrypt,
  getNetworkPrefix,
  getOnRampAssets,
  getOnRampNetworks,
  getPortalBaseURL,
  getPortalDomain,
  getPublicKeyFromSignature,
  getPublicKeyHex,
  getSHA256HashHex,
  hashPasswordWithSalt,
  initClient,
  isWalletSupported,
  mpcComputationClient,
  paraVersion,
  publicKeyFromHex,
  shortenUrl,
  toAssetInfoArray,
  retrieve as transmissionUtilsRetrieve
};
