// @getpara/user-management-client/dist/cjs/error.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var error_exports = {};
__export(error_exports, {
  ParaApiError: () => ParaApiError
});
module.exports = __toCommonJS(error_exports);
function ParaApiError(message, code, status, responseURL) {
  Error.call(this);
  if (Error.captureStackTrace) {
    Error.captureStackTrace(this, this.constructor);
  } else {
    this.stack = new Error().stack;
  }
  this.message = message;
  this.name = "ParaApiError";
  code && (this.code = code);
  status && (this.status = status);
  responseURL && (this.responseURL = responseURL);
}
const prototype = ParaApiError.prototype;
Object.defineProperty(prototype, "isParaApiError", { value: true });
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ParaApiError
});
