// @getpara/core-sdk/dist/cjs/types/config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var config_exports = {};
__export(config_exports, {
  EnabledFlow: () => EnabledFlow,
  Environment: () => Environment
});
module.exports = __toCommonJS(config_exports);
var Environment = /* @__PURE__ */ ((Environment2) => {
  Environment2["DEV"] = "DEV";
  Environment2["SANDBOX"] = "SANDBOX";
  Environment2["BETA"] = "BETA";
  Environment2["PROD"] = "PROD";
  Environment2["DEVELOPMENT"] = "BETA";
  Environment2["PRODUCTION"] = "PROD";
  return Environment2;
})(Environment || {});
var EnabledFlow = /* @__PURE__ */ ((EnabledFlow2) => {
  EnabledFlow2["BUY"] = "BUY";
  EnabledFlow2["RECEIVE"] = "RECEIVE";
  EnabledFlow2["WITHDRAW"] = "WITHDRAW";
  return EnabledFlow2;
})(EnabledFlow || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EnabledFlow,
  Environment
});
