// @getpara/core-sdk/dist/cjs/types/popup.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var popup_exports = {};
__export(popup_exports, {
  PopupType: () => PopupType
});
module.exports = __toCommonJS(popup_exports);
var PopupType = /* @__PURE__ */ ((PopupType2) => {
  PopupType2["SIGN_TRANSACTION_REVIEW"] = "SIGN_TRANSACTION_REVIEW";
  PopupType2["SIGN_MESSAGE_REVIEW"] = "SIGN_MESSAGE_REVIEW";
  PopupType2["LOGIN_PASSKEY"] = "LOGIN_PASSKEY";
  PopupType2["CREATE_PASSKEY"] = "CREATE_PASSKEY";
  PopupType2["OAUTH"] = "OAUTH";
  PopupType2["ON_RAMP_TRANSACTION"] = "ON_RAMP_TRANSACTION";
  return PopupType2;
})(PopupType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PopupType
});
