// @getpara/core-sdk/dist/cjs/types/recovery.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var recovery_exports = {};
__export(recovery_exports, {
  RecoveryStatus: () => RecoveryStatus
});
module.exports = __toCommonJS(recovery_exports);
var RecoveryStatus = /* @__PURE__ */ ((RecoveryStatus2) => {
  RecoveryStatus2["INITIATED"] = "INITIATED";
  RecoveryStatus2["READY"] = "READY";
  RecoveryStatus2["EXPIRED"] = "EXPIRED";
  RecoveryStatus2["FINISHED"] = "FINISHED";
  RecoveryStatus2["CANCELLED"] = "CANCELLED";
  return RecoveryStatus2;
})(RecoveryStatus || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RecoveryStatus
});
