// @getpara/server-sdk/dist/cjs/wallet/signing.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var signing_exports = {};
__export(signing_exports, {
  ed25519Sign: () => ed25519Sign,
  sendTransaction: () => sendTransaction,
  signMessage: () => signMessage,
  signTransaction: () => signTransaction
});
module.exports = __toCommonJS(signing_exports);
var uuid = __toESM(require("uuid"));
var import_workerWrapper = require("../workers/workerWrapper.js");
function signTransaction(ctx, userId, walletId, share, tx, chainId, sessionCookie, isDKLS) {
  return __async(this, null, function* () {
    return new Promise((resolve, reject) => __async(this, null, function* () {
      const workId = uuid.v4();
      try {
        const worker = yield (0, import_workerWrapper.setupWorker)(
          ctx,
          (sendTransactionRes) => __async(this, null, function* () {
            resolve(sendTransactionRes);
          }),
          (error) => {
            reject(error);
          },
          workId,
          {
            params: { walletId, userId, tx, chainId },
            functionType: "SIGN_TRANSACTION",
            disableWorkers: ctx.disableWorkers,
            disableWebSockets: ctx.disableWebSockets
          }
        );
        worker.postMessage({
          env: ctx.env,
          apiKey: ctx.apiKey,
          cosmosPrefix: ctx.cosmosPrefix,
          params: { share, walletId, userId, tx, chainId },
          functionType: "SIGN_TRANSACTION",
          offloadMPCComputationURL: ctx.offloadMPCComputationURL,
          disableWorkers: ctx.disableWorkers,
          sessionCookie,
          useDKLS: isDKLS,
          disableWebSockets: ctx.disableWebSockets,
          wasmOverride: ctx.wasmOverride,
          workId
        });
      } catch (error) {
        reject(error);
      }
    }));
  });
}
function sendTransaction(ctx, userId, walletId, share, tx, chainId, sessionCookie, isDKLS) {
  return __async(this, null, function* () {
    return new Promise((resolve, reject) => __async(this, null, function* () {
      const workId = uuid.v4();
      try {
        const worker = yield (0, import_workerWrapper.setupWorker)(
          ctx,
          (sendTransactionRes) => __async(this, null, function* () {
            resolve(sendTransactionRes);
          }),
          (error) => {
            reject(error);
          },
          workId,
          {
            params: { walletId, userId, tx, chainId },
            functionType: "SEND_TRANSACTION",
            disableWorkers: ctx.disableWorkers,
            disableWebSockets: ctx.disableWebSockets
          }
        );
        worker.postMessage({
          env: ctx.env,
          apiKey: ctx.apiKey,
          params: { share, walletId, userId, tx, chainId },
          functionType: "SEND_TRANSACTION",
          offloadMPCComputationURL: ctx.offloadMPCComputationURL,
          disableWorkers: ctx.disableWorkers,
          sessionCookie,
          useDKLS: isDKLS,
          disableWebSockets: ctx.disableWebSockets,
          wasmOverride: ctx.wasmOverride,
          workId
        });
      } catch (error) {
        reject(error);
      }
    }));
  });
}
function signMessage(ctx, userId, walletId, share, message, sessionCookie, isDKLS) {
  return __async(this, null, function* () {
    return new Promise((resolve, reject) => __async(this, null, function* () {
      const workId = uuid.v4();
      try {
        const worker = yield (0, import_workerWrapper.setupWorker)(
          ctx,
          (signMessageRes) => __async(this, null, function* () {
            resolve(signMessageRes);
          }),
          (error) => {
            reject(error);
          },
          workId,
          {
            params: { walletId, userId, message },
            functionType: "SIGN_MESSAGE",
            disableWorkers: ctx.disableWorkers,
            disableWebSockets: ctx.disableWebSockets
          }
        );
        worker.postMessage({
          env: ctx.env,
          apiKey: ctx.apiKey,
          params: { share, walletId, userId, message },
          functionType: "SIGN_MESSAGE",
          offloadMPCComputationURL: ctx.offloadMPCComputationURL,
          disableWorkers: ctx.disableWorkers,
          sessionCookie,
          useDKLS: isDKLS,
          disableWebSockets: ctx.disableWebSockets,
          wasmOverride: ctx.wasmOverride,
          workId
        });
      } catch (error) {
        reject(error);
      }
    }));
  });
}
function ed25519Sign(ctx, userId, walletId, share, base64Bytes, sessionCookie) {
  return __async(this, null, function* () {
    return new Promise((resolve, reject) => __async(this, null, function* () {
      const workId = uuid.v4();
      try {
        const worker = yield (0, import_workerWrapper.setupWorker)(
          ctx,
          (signMessageRes) => __async(this, null, function* () {
            resolve(signMessageRes);
          }),
          (error) => {
            reject(error);
          },
          workId,
          {
            params: { walletId, userId, base64Bytes },
            functionType: "ED25519_SIGN",
            disableWorkers: ctx.disableWorkers,
            disableWebSockets: ctx.disableWebSockets
          }
        );
        worker.postMessage({
          env: ctx.env,
          apiKey: ctx.apiKey,
          params: { share, walletId, userId, base64Bytes },
          functionType: "ED25519_SIGN",
          offloadMPCComputationURL: ctx.offloadMPCComputationURL,
          disableWorkers: ctx.disableWorkers,
          sessionCookie,
          disableWebSockets: ctx.disableWebSockets,
          wasmOverride: ctx.wasmOverride,
          workId
        });
      } catch (error) {
        reject(error);
      }
    }));
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ed25519Sign,
  sendTransaction,
  signMessage,
  signTransaction
});
