// @getpara/user-management-client/dist/cjs/types/onRamp.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var onRamp_exports = {};
__export(onRamp_exports, {
  OnRampAsset: () => OnRampAsset,
  OnRampProvider: () => OnRampProvider,
  OnRampPurchaseStatus: () => OnRampPurchaseStatus,
  OnRampPurchaseType: () => OnRampPurchaseType
});
module.exports = __toCommonJS(onRamp_exports);
var OnRampProvider = /* @__PURE__ */ ((OnRampProvider2) => {
  OnRampProvider2["RAMP"] = "RAMP";
  OnRampProvider2["STRIPE"] = "STRIPE";
  OnRampProvider2["MOONPAY"] = "MOONPAY";
  return OnRampProvider2;
})(OnRampProvider || {});
var OnRampAsset = /* @__PURE__ */ ((OnRampAsset2) => {
  OnRampAsset2["ETHEREUM"] = "ETHEREUM";
  OnRampAsset2["USDC"] = "USDC";
  OnRampAsset2["TETHER"] = "TETHER";
  OnRampAsset2["POLYGON"] = "POLYGON";
  OnRampAsset2["SOLANA"] = "SOLANA";
  OnRampAsset2["ATOM"] = "ATOM";
  OnRampAsset2["CELO"] = "CELO";
  OnRampAsset2["CUSD"] = "CUSD";
  OnRampAsset2["CEUR"] = "CEUR";
  OnRampAsset2["CREAL"] = "CREAL";
  OnRampAsset2["BERA"] = "BERA";
  return OnRampAsset2;
})(OnRampAsset || {});
var OnRampPurchaseStatus = /* @__PURE__ */ ((OnRampPurchaseStatus2) => {
  OnRampPurchaseStatus2["INITIATED"] = "INITIATED";
  OnRampPurchaseStatus2["FINISHED"] = "FINISHED";
  OnRampPurchaseStatus2["CANCELLED"] = "CANCELLED";
  return OnRampPurchaseStatus2;
})(OnRampPurchaseStatus || {});
var OnRampPurchaseType = /* @__PURE__ */ ((OnRampPurchaseType2) => {
  OnRampPurchaseType2["BUY"] = "BUY";
  OnRampPurchaseType2["SELL"] = "SELL";
  return OnRampPurchaseType2;
})(OnRampPurchaseType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OnRampAsset,
  OnRampProvider,
  OnRampPurchaseStatus,
  OnRampPurchaseType
});
